webapp/src/pages/ChatroomPage/ChatroomPage.tsx

```diff
@@ -1,42 +1,104 @@
 import { ChatroomMessageInput, ChatroomMessages } from './components';
-import { Header, IconVariant } from '../../shared/components';
+import {
+  Header,
+  IconVariant,
+  Loading,
+  Text,
+  TextColor,
+  TextVariant,
+} from '../../shared/components';
+import { CHATROOM_URL } from '../../shared/urls';
 import socket from '../../shared/socket';
 import './ChatroomPage.css';
 import { useNavigation } from '../../shared/hooks/UseNavigation';
-import { WsException } from '../../shared/types';
+import { SubmitStatus, WsException } from '../../shared/types';
+import { useData } from '../../shared/hooks/UseData';
+import { chatApi } from '../../shared/services/ApiService';
+import NotFoundPage from '../NotFoundPage/NotFoundPage';
+import { useAuth } from '../../shared/hooks/UseAuth';
+import { User } from '../../shared/generated';
 
-function Chatroom() {
+export default function ChatroomPage() {
   const { chatroomId } = useParams();
+  const { authUser } = useAuth();
+
+  if (!chatroomId || !authUser) {
+    return null;
+  }
+  return <Chatroom chatroomId={chatroomId} authUser={authUser} />;
+}
+
+type ChatroomProps = {
+  chatroomId: string;
+  authUser: User;
+};
+
+function Chatroom({ chatroomId, authUser }: ChatroomProps) {
   const { goBack } = useNavigation();
+  const [status, setStatus] = useState<SubmitStatus>({
+    type: 'pending',
+    message: '',
+  });
+  const getChatroom = useCallback(
+    () => chatApi.chatControllerGetChatroomById({ id: chatroomId }),
+    [chatroomId],
+  );
+  const { data: chatroom, isLoading: isChatroomLoading } = useData(getChatroom);
+  const getChatroomMember = useCallback(
+    () =>
+      chatApi.chatControllerGetChatroomMember({
+        chatroomId,
+        userId: authUser.id,
+      }),
+    [chatroomId, authUser.id],
+  );
+  const { data: chatroomMember, isLoading: isChatroomMemberLoading } =
+    useData(getChatroomMember);
 
   useEffect(() => {
     socket.on('exception', (wsError: WsException) => {
-      // TODO Add error notification in case of an WebSocket exception
-      console.log(wsError);
+      setStatus({ type: 'error', message: wsError.message });
     });
-    if (chatroomId) {
-      socket.emit('joinChatroom', { chatroomId });
-    }
+    socket.emit('joinChatroom', { chatroomId });
 
     return () => {
       socket.off('exception');
-      if (chatroomId) {
-        socket.emit('leaveChatroom', { chatroomId });
-      }
+      socket.emit('leaveChatroom', { chatroomId });
     };
   }, [chatroomId]);
 
-  if (!chatroomId) {
-    return null;
+  useEffect(() => {
+    const timer = setTimeout(() => {
+      setStatus({ type: 'pending', message: '' });
+    }, 5000);
+    return () => clearTimeout(timer);
+  }, [status]);
+
+  if (isChatroomLoading || isChatroomMemberLoading) {
+    return (
+      <div className="chatroom">
+        <div className="chatroom-loading">
+          <Loading />
+        </div>
+      </div>
+    );
+  }
+
```

Comment:
This way we will always show the loading spinner, even for an instant. I don't now if it is a good practice, but what do you think about fooling the eye a little bit loading the immediate components of the page and the ones with a query a bit later.
I mean, for example, for the chatroom name, removing this `Loading` component and in line 101 something like this `{isChatroomLoading ? '' : chatroom!.name} `. But I insist, I don't know if this make sense or is a complete nuisance.

Reply:
I see your point, and I'm not sure which is better for this page.
If it takes a long time to fetch the chatroom info because the user has slow internet or the server is under heavy load, the user will see an empty chatroom, and if they try to send messages, it won't work.
On the other side, they will see a loading spinner.

Also, code-wise will be pretty hard to understand what and when it renders because we have to add some conditionals.

We could display a skeleton screen on pages that don't have much user interactivity, like viewing a user profile page, chatroom details page, etc.

What do you think?

Reply:
Agree with @gbudau , the alternative would be using skeletons for those components not loaded yet but I think that would be overkill with the time we have.

Managing skeletons is not easy (you have to predict content size so user don't see unexpected jumps, we would have to rework some pages to make the more modular etc.)

I'd say a good enough solution is what we have, a simple loader for the whole screen. That being said, thanks for bringing the point, I think it's an interesting one :+1: